keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -3,703 +3,644 @@
-    rpc ProvisionerRevokeBucketAccess (ProvisionerRevokeBucketAccessRequest) returns (ProvisionerRevokeBucketAccessResponse);
-}
 ```
 
-## ProvisionerGetInfo
+The API of `bucket.json` follows the same versioning and lifecycle as the rest of Bucket* APIs. Workloads are expected to read the definition in this file to access a bucket.
 
-This call is meant to retrieve the unique provisioner Identity. This identity will have to be set in `BucketRequest.Provisioner` field in order to invoke this specific provisioner.
+# COSI API Reference
 
-```
-message ProvisionerGetInfoRequest {
-    // Intentionally left blank
-}
+## Bucket 
 
-message ProvisionerGetInfoResponse {
-  // The name MUST follow domain name notation format
-  // (https://tools.ietf.org/html/rfc1035#section-2.3.1). It SHOULD
-  // include the plugin's host company name and the plugin name,
-  // to minimize the possibility of collisions. It MUST be 63
-  // characters or less, beginning and ending with an alphanumeric
-  // character ([a-z0-9A-Z]) with dashes (-), dots (.), and
-  // alphanumerics between. This field is REQUIRED.
-  string name = 1;
+Resource to represent a Bucket in OSP. Buckets are cluster scoped.
+
+```yaml
+Bucket {
+  TypeMeta
+  ObjectMeta
+  
+  Spec BucketSpec {
+    // Provisioner is the name of driver associated with this bucket
+    Provisioner string 
+
+    // DeletionPolicy is used to specify how COSI should handle deletion of this 
+    // bucket. There are 3 possible values:
+    //  - Retain: Indicates that the bucket should not be deleted from the OSP
+    //  - Delete: Indicates that the bucket should be deleted from the OSP
+    //        once all the workloads accessing this bucket are done
+    //  - ForceDelete: Indicates that the bucket and its contents should be 
+    //        deleted from the OSP even if workloads are currently accessing 
+    //        this bucket
+    DeletionPolicy DeletionPolicy 
+    
+    // Name of the BucketClass specified in the BucketRequest
+    // +optional
+    BucketClassName string 
+    
+    // Name of the BucketRequest that resulted in the creation of this Bucket
+    // +optional
+    BucketRequest corev1.ObjectReference
+    
+    // AllowedNamespaces is the list of namespaces from which workloads  
+    // can access this Bucket. If empty, all namespaces can access this access
+    // +optional 
+    AllowedNamespaces []string 
+
+    // Protocol is the data API this bucket is expected to adhere to.
+    // The possible values for protocol are:
+    // -  S3: Indicates Amazon S3 protocol
+    // -  Azure: Indicates Microsoft Azure BlobStore protocol
+    // -  GCS: Indicates Google Cloud Storage protocol
+    Protocol Protocol
+    
+    // Parameters is an opaque map for passing in configuration to a driver
+    // for creating the bucket
+    // +optional
+    Parameters map[string]string 
+  }    
```

Comment:
There is no API linkage between `Bucket` and `BucketAccess`. So will COSI be making GET (or cache) calls to fetch all BAs for a given B?

Reply:
Also, I haven't yet seen 1:n mapping relations defined between the COSI resources. Is it 1 BR per B? One BAR per BA? Many BAs per B? etc.

Reply:
During deletion is the only time we need this list. In which case, this cascading delete like behavior, where deleting a BR deletes the B, and all other associated  BA's can be achieved using owner references. 

Is there any other reason for us to get this information? If not, then we should be fine

Reply:
> Also, I haven't yet seen 1:n mapping relations defined between the COSI resources. Is it 1 BR per B? One BAR per BA? Many BAs per B? etc.

The resources are loosely coupled, for instance, similar to how pods and nodes are related. Whenever a stronger coupling is needed, we directly reference the related object in the referring object.